### PR TITLE
LinearFormExtension fixes

### PR DESCRIPTION
**Pull request by pazner**
- Adds common abstract base class for ElementRestriction operators
    - Adds `AddMultTranspose` to interface
- Use `DETERMINANTS` instead of `JACOBIANS` in DomainLFIntegrator
- Properly accumulate integrators in LinearFormExtension
- Add test for accumulation of integrators in LinearFormExtension

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3143](https://github.com/mfem/mfem/pull/3143) | @pazner | @tzanio | @psocratis + @camierjs + @YohannDudouit | 8/16/22 | 9/26/22 | 9/29/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: fem/lininteg_domain.cpp [hunk 1]
```diff
@@ -25,7 +25,7 @@ void DLFEvalAssemble2D(const int vdim, const int ne, const int d, const int q,
    const auto F = coeff.Read();
    const auto M = Reshape(markers, ne);
    const auto B = Reshape(b, q, d);
-   const auto J = Reshape(j, q, q, 2,2, ne);
+   const auto J = Reshape(j, q, q, ne);
```
Comment (camierjs) [inline fem/lininteg_domain.cpp:28]: Could we change `J` to `D` maybe?

Comment (pazner) [reply fem/lininteg_domain.cpp:28]: How about `DETJ`? `D` is already used for `constexpr int D = T_D1D ? T_D1D : MAX_D1D;`

Comment (pazner) [reply fem/lininteg_domain.cpp]: Done in 106964e

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by camierjs — APPROVED** [review]

**Review by YohannDudouit — CHANGES_REQUESTED** [review]

File: fem/linearform_ext.cpp [hunk 2]
```diff
@@ -79,7 +80,8 @@ void LinearFormExtension::Update()
    for (int i = 0; i < NE; ++i) { attributes[i] = mesh.GetAttribute(i); }
 
    constexpr ElementDofOrdering ordering = ElementDofOrdering::LEXICOGRAPHIC;
-   elem_restrict_lex = fes.GetElementRestriction(ordering);
+   elem_restrict_lex = dynamic_cast<const ElementRestrictionOperator*>(
```
[the remark this answers is not in the document]
Comment (pazner) [reply fem/linearform_ext.cpp]: Good idea, do the changes in e03a7fef417fb11a5337c26ed9492082b4fa8406 look OK to you?

File: fem/restriction.hpp [hunk 3]
```diff
@@ -85,14 +95,19 @@ class ElementRestriction : public Operator
    /** Fill the J and Data arrays of SparseMatrix corresponding to the sparsity
        pattern given by this ElementRestriction, and the values of ea_data. */
    void FillJAndData(const Vector &ea_data, SparseMatrix &mat) const;
+   /// @private Not part of the public interface (device kernel limitation).
+   ///
+   /// Performs either MultTranpose or AddMultTranposes depending on the boolean
```
Comment (psocratis) [inline fem/restriction.hpp:100]: typo: `MultTranspose`, `AddMultTranspose`

File: fem/restriction.hpp [hunk 4]
```diff
@@ -103,12 +118,18 @@ class L2ElementRestriction : public Operator
...
+   void AddMultTranspose(const Vector &x, Vector &y) const;
    /** Fill the I array of SparseMatrix corresponding to the sparsity pattern
        given by this ElementRestriction. */
    void FillI(SparseMatrix &mat) const;
    /** Fill the J and Data arrays of SparseMatrix corresponding to the sparsity
        pattern given by this L2FaceRestriction, and the values of ea_data. */
    void FillJAndData(const Vector &ea_data, SparseMatrix &mat) const;
+   /// @private Not part of the public interface (device kernel limitation).
+   ///
+   /// Performs either MultTranpose or AddMultTranposes depending on the boolean
```
Comment (psocratis) [inline fem/restriction.hpp:130]: typo: `MultTranspose`, `AddMultTranspose`

File: fem/lininteg_domain.cpp [hunk 5]
```diff
@@ -260,7 +229,7 @@ static void DLFEvalAssemble(const FiniteElementSpace &fes,
    const int ne = fes.GetMesh()->GetNE();
    const int *M = markers.Read();
    const double *B = maps.B.Read();
-   const double *J = geom->J.Read();
+   const double *J = geom->detJ.Read();
```
Comment (psocratis) [inline fem/lininteg_domain.cpp:232]: maybe we name this `detJ` as well, just for consistency?

**Review by psocratis — APPROVED** [review]
a couple of typos .... otherwise it looks good to me

File: fem/restriction.hpp (hunk 3 above)
Comment (pazner) [reply fem/restriction.hpp]: Fixed in 6b38f5c46c96a097d28d374cf5e0fb2ab002e4d5

File: fem/restriction.hpp (hunk 4 above)
Comment (pazner) [reply fem/restriction.hpp]: Fixed in 6b38f5c46c96a097d28d374cf5e0fb2ab002e4d5

File: fem/lininteg_domain.cpp (hunk 5 above)
Comment (pazner) [reply fem/lininteg_domain.cpp]: Fixed in 1f18293a2f55152c15153084045d9b0795079870

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Should we mention some of the changes in `CHANGELOG`?

**Pull request merged**